### PR TITLE
KAFKA-14133: Move StreamsMetricsImplTest to Mockito 

### PR DESCRIPTION
**Pull request by clolov**
This pull requests migrates mocks from StreamsMetricsImplTest to Mockito.

### CODE REVIEW

**Comment by clolov** [comment]
~Local run of the failing tests passes:~
```
> Configure project :
Starting build with version 3.7.0-SNAPSHOT (commit id 79afb498) using Gradle 8.3, Java 1.8 and Scala 2.13.12
Build properties: maxParallelForks=10, maxScalacThreads=8, maxTestRetries=0
...
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldReturnSameMetricIfAlreadyCreated PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetExistingNodeLevelSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetCacheLevelTagMap PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > testThroughputMetrics PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldRemoveStateStoreLevelSensors PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddMinAndMaxMetricsToSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > testTotalMetricDoesntDecrease PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddSum PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldNotUseSameStoreLevelSensorKeyWithDifferentThreadIds PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddAvgAndTotalMetricsToSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldCleanupThreadLevelImmutableMetric PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > testLatencyMetrics PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddThreadLevelMutableMetric PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddClientLevelImmutableMetric PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetExistingClientLevelSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddNewStoreLevelMutableMetric PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetNewTaskLevelSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddRateTotalSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetNewClientLevelSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetNewStoreLevelSensorIfNoneExists PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddAmountRateAndSum PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetNewNodeLevelSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldMeasureLatency PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetExistingStoreLevelSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetClientLevelTagMap PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > testMultiLevelSensorRemoval PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddClientLevelMutableMetric PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddThreadLevelImmutableMetric PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddInvocationRateToSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldThrowIfRateTotalSensorIsAddedWithOddTags PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > testRemoveSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetExistingThreadLevelSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddAvgAndMinAndMaxMetricsToSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldCreateMetricOnceDuringConcurrentMetricCreationRequest PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetNewCacheLevelSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetNewThreadLevelSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldUseSameStoreLevelSensorKeyWithSameSensorNames PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetExistingCacheLevelSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetStoreLevelTagMap PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddRateTotalSensorWithCustomTags PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldRemoveClientLevelMetricsAndSensors PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetThreadLevelTagMap PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddTotalCountAndSumMetricsToSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddLatencyRateTotalSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddAmountRate PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetNewTopicLevelSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldNotMeasureLatencyDueToRecordingLevel PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldNotAddStoreLevelMutableMetricIfAlreadyExists PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldProvideCorrectStrings PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldThrowIfLatencyRateTotalSensorIsAddedWithOddTags PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldNotUseSameStoreLevelSensorKeyWithDifferentTaskIds PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldReturnMetricsVersionCurrent PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldNotMeasureLatencyBecauseSensorHasNoMetrics PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldNotUseSameStoreLevelSensorKeyWithDifferentStoreNames PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddLatencyRateTotalSensorWithCustomTags PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetExistingTopicLevelSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetExistingTaskLevelSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddValue PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > testRemoveNullSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldCreateNewStoreLevelMutableMetric PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > testNullMetrics PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldUseSameStoreLevelSensorKeyWithTwoDifferentSensorNames PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldCleanupThreadLevelMutableMetric PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldRemoveThreadLevelSensors PASSED
...
BUILD SUCCESSFUL in 2s
```

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImplTest.java
Comment (divijvaidya) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImplTest.java:1237]: for my curiosity, does strict subs verify usage of stubs defined using `doNothin()` as well or just for subs defined using `when()`

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImplTest.java
Comment (divijvaidya) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImplTest.java]: This we are not printing details when assertion fails, can you please add a message to print the thatMetricConfig similar to `assertTrue(eqMetricConfig(expected, actual), "my message here")`

Will help in debugging failed tests and maintain parity with existing code. Maybe someone added such detailed logging in one place because debugging this was very hard!

**Comment by clolov** [comment]
I am able to reproduce the failing tests locally and I am trying to understand why the behaviour is not as expected.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImplTest.java
Comment (clolov) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImplTest.java]: Okay, makes sense!

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImplTest.java
Comment (clolov) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImplTest.java:1237]: I will aim to double-check, but I was operating under this assumption!

Comment (clolov) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImplTest.java:1237]: If I add `doNothing().when(sensor).checkQuotas();` in this test I get the following error when running the whole test suite:
```
org.mockito.exceptions.misusing.UnnecessaryStubbingException: 
Unnecessary stubbings detected in test class: StreamsMetricsImplTest
Clean & maintainable test code requires zero unnecessary code.
Following stubbings are unnecessary (click to navigate to relevant line of code):
  1. -> at org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsImplTest.shouldMeasureLatency(StreamsMetricsImplTest.java:1209)
```
As such, I believe this means that the verification is carried out on these stubs as well.

File: gradle/dependencies.gradle
Comment (ijuma) [inline gradle/dependencies.gradle]: This comment is not accurate, `mockito-core` from version `5` does support that. The issue is that Scala 2.12 uses mockito 4. So, the additional check makes sense, but the comment doesn't.

Comment (clolov) [reply gradle/dependencies.gradle]: Yes, you are absolutely correct. Unfortunately, even with the additional check I still cannot quite make all the tests to pass, so I will continue battling this. But thank you for the review!

Comment (ijuma) [reply gradle/dependencies.gradle]: Yes, we don't run Scala 2.12 with Java 11+ in CI so it would not impact the PR builds.

**Comment by clolov** [comment]
No failures in `StreamsMetricsImplTest` in the last build! I will just make the final changes based on comments and this should be good to go.

**Comment by clolov** [comment]
Heya @ijuma and @divijvaidya, none of the test failures are related to the test being changed in this pull request and the build passes! I hope I have addressed all of your comments and this is ready to merge but if not, let me know what else I can do!

**Review by divijvaidya — APPROVED** [review]
Thanks Christo. The changes look good to me.

**Comment by divijvaidya** [comment]
Test failures are unrelated:
```
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.clients.consumer.CooperativeStickyAssignorTest.testLargeAssignmentAndGroupWithNonEqualSubscription(boolean).hasConsumerRack = false](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14623/6/testReport/junit/org.apache.kafka.clients.consumer/CooperativeStickyAssignorTest/Build___JDK_11_and_Scala_2_13___testLargeAssignmentAndGroupWithNonEqualSubscription_boolean__hasConsumerRack___false/)
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.clients.consumer.StickyAssignorTest.testLargeAssignmentAndGroupWithNonEqualSubscription(boolean).hasConsumerRack = false](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14623/6/testReport/junit/org.apache.kafka.clients.consumer/StickyAssignorTest/Build___JDK_11_and_Scala_2_13___testLargeAssignmentAndGroupWithNonEqualSubscription_boolean__hasConsumerRack___false/)
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.controller.QuorumControllerTest.testFenceMultipleBrokers()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14623/6/testReport/junit/org.apache.kafka.controller/QuorumControllerTest/Build___JDK_11_and_Scala_2_13___testFenceMultipleBrokers__/)
[Build / JDK 17 and Scala 2.13 / org.apache.kafka.streams.integration.StoreQueryIntegrationTest.shouldQuerySpecificStalePartitionStoresMultiStreamThreads()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14623/6/testReport/junit/org.apache.kafka.streams.integration/StoreQueryIntegrationTest/Build___JDK_17_and_Scala_2_13___shouldQuerySpecificStalePartitionStoresMultiStreamThreads__/)
[Build / JDK 8 and Scala 2.12 / kafka.server.ReplicaManagerTest.[1] enableRemoteStorage=true](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14623/6/testReport/junit/kafka.server/ReplicaManagerTest/Build___JDK_8_and_Scala_2_12____1__enableRemoteStorage_true/)
[Build / JDK 8 and Scala 2.12 / kafka.server.ReplicaManagerTest.[1] enableRemoteStorage=true](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14623/6/testReport/junit/kafka.server/ReplicaManagerTest/Build___JDK_8_and_Scala_2_12____1__enableRemoteStorage_true_2/)
[Build / JDK 8 and Scala 2.12 / kafka.zk.ZkMigrationIntegrationTest.[4] Type=ZK, MetadataVersion=3.7-IV0, Security=PLAINTEXT](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14623/6/testReport/junit/kafka.zk/ZkMigrationIntegrationTest/Build___JDK_8_and_Scala_2_12____4__Type_ZK__MetadataVersion_3_7_IV0__Security_PLAINTEXT/)
[Build / JDK 21 and Scala 2.13 / integration.kafka.server.FetchFromFollowerIntegrationTest.testRackAwareRangeAssignor(String).quorum=zk](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14623/6/testReport/junit/integration.kafka.server/FetchFromFollowerIntegrationTest/Build___JDK_21_and_Scala_2_13___testRackAwareRangeAssignor_String__quorum_zk/)
[Build / JDK 21 and Scala 2.13 / org.apache.kafka.streams.integration.ConsistencyVectorIntegrationTest.shouldHaveSamePositionBoundActiveAndStandBy](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14623/6/testReport/junit/org.apache.kafka.streams.integration/ConsistencyVectorIntegrationTest/Build___JDK_21_and_Scala_2_13___shouldHaveSamePositionBoundActiveAndStandBy/)
[Build / JDK 21 and Scala 2.13 / org.apache.kafka.streams.integration.RestoreIntegrationTest.shouldInvokeUserDefinedGlobalStateRestoreListener()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14623/6/testReport/junit/org.apache.kafka.streams.integration/RestoreIntegrationTest/Build___JDK_21_and_Scala_2_13___shouldInvokeUserDefinedGlobalStateRestoreListener__/)
```

**Comment by divijvaidya** [comment]
@ijuma I am merging this in without waiting for your approval since the changes are straightforward and your comments have been addressed.  Please leave a comment if something is missing and we can follow up in a separate PR.

**Comment by ijuma** [comment]
Sounds good, thanks.

**Pull request merged**
